upup/pkg/fi/cloudup/awstasks/launchtemplate.go

```diff
@@ -0,0 +1,526 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awstasks
+
+import (
+	"encoding/base64"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/awsup"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/golang/glog"
+)
+
+// LaunchTemplate defines the specificate for a template
+type LaunchTemplate struct {
+	// Name is the name of the configuration
+	Name *string
+	// Lifecycle is the resource lifecycle
+	Lifecycle *fi.Lifecycle
+
+	// AssociatePublicIP indicates if a public ip address is assigned to instabces
+	AssociatePublicIP *bool
+	// IAMInstanceProfile is the IAM profile to assign to the nodes
+	IAMInstanceProfile *IAMInstanceProfile
+	// ID is the launch configuration name
+	ID *string
+	// ImageID is the AMI to use for the instances
+	ImageID *string
+	// InstanceMonitoring indicates if monitoring is enabled
+	InstanceMonitoring *bool
+	// InstanceType is the machine type to use
+	InstanceType *string
+	// If volume type is io1, then we need to specify the number of Iops.
+	RootVolumeIops *int64
+	// RootVolumeOptimization enables EBS optimization for an instance
+	RootVolumeOptimization *bool
+	// RootVolumeSize is the size of the EBS root volume to use, in GB
+	RootVolumeSize *int64
+	// RootVolumeType is the type of the EBS root volume to use (e.g. gp2)
+	RootVolumeType *string
+	// SSHKey is the ssh key for the instances
+	SSHKey *SSHKey
+	// SecurityGroups is a list of security group associated
+	SecurityGroups []*SecurityGroup
+	// SpotPrice is set to the spot-price bid if this is a spot pricing request
+	SpotPrice string
+	// Tenancy. Can be either default or dedicated.
+	Tenancy *string
+	// UserData is the user data configuration
+	UserData *fi.ResourceHolder
+}
+
+var (
+	_ fi.CompareWithID     = &LaunchTemplate{}
+	_ fi.ProducesDeletions = &LaunchTemplate{}
+	_ fi.Deletion          = &deleteLaunchTemplate{}
+)
+
+// CompareWithID implements the comparable interface
+func (t *LaunchTemplate) CompareWithID() *string {
+	return t.ID
+}
+
+// LaunchTemplateName returns the lanuch template name
+func (t *LaunchTemplate) LaunchTemplateName() string {
+	return fmt.Sprintf("%s-%s", fi.StringValue(t.Name), fi.BuildTimestampString())
```

Comment:
I see this is using similar behavior as launch configurations in that we're suffixing them with a timestamp and sorting those to determine the latest version to use. Since [LaunchTemplates support versioning](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeLaunchTemplateVersions.html) I wonder if we should drop the timestamp suffix from the LT name and instead rely on using the latest or "default" version of the LT itself. This would result in less launch template resources being created in AWS which would probably be preferable.

It may complicate the task name or ID, as well as the aws_cloud.go logic used for determining out of date instances in a rolling-update.

Reply:
Makes sense ...I'll take a look and see if we can switch to using the template versions